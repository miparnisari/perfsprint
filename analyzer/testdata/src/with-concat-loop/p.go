package p

func positive() {
	var s string
	words := []string{"one", "two", "three"}
	for w := range words {
		s += words[w] // want "concat-loop: string concatenation in a loop"
	}
	for w := range words {
		s = s + words[w] // want "concat-loop: string concatenation in a loop"
	}
	for w := 0; w < 10; w++ {
		s = s + "y" // want "concat-loop: string concatenation in a loop"
	}
	for w := 0; w < 10; w++ {
		if w%2 == 1 {
			s = s + "y" // want "concat-loop: string concatenation in a loop"
		}
	}
	nb := 0
	for w := 0; w < 10; w++ {
		if w%2 == 1 {
			nb += 1
		} else {
			s = s + "y" // want "concat-loop: string concatenation in a loop"
		}
	}
}

func negative() {
	for w := 0; w < 10; w++ {
		s := "local"
		s = s + "y"
	}
	for w := 0; w < 10; w++ {
		var s string
		s = s + "y"
	}
	for w := 0; w < 10; w++ {
		var s2, s string
		s = s + "y"
		_ = s
		_ = s2
	}
	for w := 0; w < 10; w++ {
		s2, s := "local", "same"
		s = s + "y"
		_ = s
		_ = s2
	}
	nb := 0
	for w := 0; w < 10; w++ {
		nb += w
	}
	for w := 0; w < 10; w++ {
		nb = nb + w
	}
	words := []string{"one", "two", "three"}
	var s string
	for w := range words {
		s = "toto" + words[w]
	}
	var s2 string
	for w := range words {
		s = s2 + words[w]
	}
	_ = s
	_ = s2
}
