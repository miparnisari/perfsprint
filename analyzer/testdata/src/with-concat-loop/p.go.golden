package p

func positive() {
	var s string
	words := []string{"one", "two", "three"}
	var sSb6 strings.Builder
	for w := range words {
		sSb6.WriteString(words[w]) // want "concat-loop: string concatenation in a loop"
	}
	s += sSb6.String()
	var sSb9 strings.Builder
	for w := range words {
		sSb9.WriteString(words[w]) // want "concat-loop: string concatenation in a loop"
	}
	s += sSb9.String()
	var sSb12 strings.Builder
	for w := 0; w < 10; w++ {
		sSb12.WriteString("y") // want "concat-loop: string concatenation in a loop"
	}
	s += sSb12.String()
	var sSb15 strings.Builder
	for w := 0; w < 10; w++ {
		if w%2 == 1 {
			sSb15.WriteString("y") // want "concat-loop: string concatenation in a loop"
		}
	}
	s += sSb15.String()
	nb := 0
	var sSb21 strings.Builder
	for w := 0; w < 10; w++ {
		if w%2 == 1 {
			nb += 1
		} else {
			sSb21.WriteString("y") // want "concat-loop: string concatenation in a loop"
		}
	}
	s += sSb21.String()
	var sSb28 strings.Builder
	for w := 0; w < 10; w++ {
		if w%2 == 1 {
			sSb28.WriteString("x") // want "concat-loop: string concatenation in a loop"
		} else {
			sSb28.WriteString("y")
		}
	}
	s += sSb28.String()
	s2 := "prefix"
	var sSb36 strings.Builder
	var s2Sb36 strings.Builder
	for w := 0; w < 10; w++ {
		if w%2 == 1 {
			s2Sb36.WriteString("x")
		} else {
			sSb36.WriteString("y") // want "concat-loop: string concatenation in a loop"
		}
	}
	s += sSb36.String()
	s2 += s2Sb36.String()
}

func negative() {
	for w := 0; w < 10; w++ {
		s := "local"
		s = s + "y"
	}
	for w := 0; w < 10; w++ {
		var s string
		s = s + "y"
	}
	for w := 0; w < 10; w++ {
		var s2, s string
		s = s + "y"
		_ = s
		_ = s2
	}
	for w := 0; w < 10; w++ {
		s2, s := "local", "same"
		s = s + "y"
		_ = s
		_ = s2
	}
	nb := 0
	for w := 0; w < 10; w++ {
		nb += w
	}
	for w := 0; w < 10; w++ {
		nb = nb + w
	}
	words := []string{"one", "two", "three"}
	var s string
	for w := range words {
		s = "toto" + words[w]
	}
	var s2 string
	for w := range words {
		s = s2 + words[w]
	}
	_ = s
	_ = s2
}
